Readme.md
# Grafterm [![CircleCI][circleci-image]][circleci-url] [![Go Report Card][go-reportcard-image]][go-reportcard-url]

Visualize metrics dashboards on the terminal, like a simplified and minimalist version of [Grafana] for terminal.

![grafterm red dashboard](/img/grafterm-red-compressed.gif)

## Features

- Multiple widgets (graph, singlestat, gauge).
- Multiple datasources usage.
- User stored datasources.
- Override dashboard datasource ID to different datasource ID configured by the user.
- Custom dashboards based on JSON configuration files.
- Extensible metrics datasource implementation (Prometheus and Graphite included).
- Templating of variables.
- Auto time interval adjustment for queries.
- Auto unit formatting on widgets.
- Fixed and adaptive grid.
- Color customization on widgets.
- Configurable autorefresh.
- Single binary and easy usage/deployment.

## Installation

Download the binaries from [releases]

## Running options

Exit with `q` or `Esc`

### Simple

```bash
grafterm -c ./mydashboard.json
```

### Relative time

```bash
grafterm -c ./mydashboard.json -d 48h
```

### Refresh interval

```bash
grafterm -c ./mydashboard.json -r 2s
```

### Debugging

When grafterm doesn't show anything may be that has errors getting metrics or similar. There is available a `--debug` flag that will write a log on `grafterm.log` (this path can be override with `--log-path` flag)

Read the log

```bash
tail -f ./grafterm.log
```

And run grafterm in debug mode.

```bash
grafterm -c ./mydashboard.json  -d 48h -r 2s --debug
```

### Fixed time

Setting a fixed time range to visualize the metrics using duration notation. In this example is start at `now-22h` and end at `now-20h`

```bash
grafterm -c ./mydashboard.json -s 22h -e 20h
```

Setting a fixed time range to visualize the metrics using timestamp [ISO 8601] notation.

```bash
grafterm -c ./mydashboard.json -s 2019-05-12T12:32:11+02:00 -e 2019-05-12T12:35:11+02:00
```

### Replacing dashboard variables

```bash
grafterm -c ./mydashboard.json -v env=prod -v job=envoy
```

### Replacing dashboard datasource configuration

Replace dashbaord `prometheus` datasource with user datasource `thanos-prometheus` (check [Datasources](#datasources) section):

```bash
grafterm -c ./mydashboard.json -a "prometheus=thanos-prometheus"
```

Replace dashboard `prometheus` datasource with user datasource `thanos-prometheus` available on `/tmp/my-datasources.json` user datasource configuration file:

```bash
grafterm -c ./mydashboard.json -a "prometheus=thanos-prometheus" -u /tmp/my-datasources.json
```

## Dashboard

Check [this][cfg-md] section that explains how a dashboard is configured. Also check [dashboard examples][dashboard-examples]

## Datasources

Datasources are the way grafterm knows how to retrieve the metrics for the dashboard.

check available types and how to configure in [this][cfg-md] section.

**If you want support for a new datasource type, open an issue or send a PR**

### Overriding dashboard datasources

Dashboard referenced datasources on the queries can be override.

#### User datasource

Grafterm dashboards can have default datasources but the user can override these datasources using a datasources config file. This file has the same format as the dashboard configuration file but will ignore anything other than the `datasources` block. Example:

```json
{
  "version": "v1",
  "datasources": {
    "prometheus": {
      "prometheus": { "address": "http://127.0.0.1:9090" }
    },
    "localprom": {
      "prometheus": { "address": "http://127.0.0.1:9091" }
    },
    "thanos": {
      "prometheus": { "address": "http://127.0.0.1:9092" }
    },
    "m3db": {
      "prometheus": { "address": "http://127.0.0.1:9093" }
    },
    "victoriametrics": {
      "prometheus": { "address": "http://127.0.0.1:8428" }
    },
    "wikimedia": {
      "graphite": { "address": "https://graphite.wikimedia.org" }
    }
  }
}
```

If the dashboard has defined a datasource configuration with the ID `my-ds` reference, and the user datasources has this same datasource ID, grafterm will use the user defined one when the queries in the dashboard reference this ID.

The user datasources location can be configured with this priority (from highest to lowest):

- If `--user-datasources` explicit flag is used, it will use this.
- If `GRAFTERM_USER_DATASOURCES` env var is set, it will use this.
- As a fallback location will check `{USER_HOME}/grafterm/datasources.json` exists.

#### Alias

Apart from overriding the dashboard datasources IDs that match with the user datasources, the user can force an alias with the form `dashboard-ds-id=user-ds-id`.

For example, the dashboard uses a datasource named `prometheus-2b`, and we want to use our local prometheus configured on the user datasources as `localprom`, we could use the alias flag like this: `-a "prometheus-2b=localprom"`, now every query the dashboard widgets make to `prometheus-2b` will be made to `localprom`.

## Kudos

This project would not be possible without the effort of many people and projects but specially [Grafana] for the inspiration, ideas and the project itself, and [Termdash] for the rendering of all those fancy graphs on the terminal.

[circleci-image]: https://img.shields.io/circleci/project/github/slok/grafterm/master.svg
[circleci-url]: https://circleci.com/gh/slok/grafterm
[go-reportcard-image]: https://goreportcard.com/badge/github.com/slok/grafterm
[go-reportcard-url]: https://goreportcard.com/report/github.com/slok/grafterm
[grafana]: https://grafana.com/
[termdash]: https://github.com/mum4k/termdash
[releases]: https://github.com/slok/grafterm/releases
[cfg-md]: /docs/cfg.md
[dashboard-examples]: /dashboard-examples
[iso 8601]: https://en.wikipedia.org/wiki/ISO_8601
[prometheus]: http://prometheus.io


CHANGELOG.md
# Changelog

## [0.2.1] - 2021-05-07

### Fixed

- Updated Testify to v1.7.0 to resovle issue #60; a CVE 6.5 present in the Yaml package that Testify v1.4.0 uses.

## [0.2.0] - 2019-07-26

### Added

- ARM release builds.
- Add `nullPointMode` series override setting (with `zero` and `connected` strategies).
- Graphite datasource.
- Milliseconds unit conversion.
- Quit grafterm with `Esc` key.
- User defined datasources via flag and/or env var.
- Alias flag to override dashboard datasource ID using user datasource IDs.
- Fallback dashboard referenced datasources to user datasources.

### Fixed

- Gauges that had color thresholds not being show.

## [0.1.0] - 2019-05-13

### Added

- `start` and `end` flags to visualize fixed time range graphs.
- `var` repeatable flag to override dashboard variables.
- Unit formatters for time, RPS, percent and ratios.
- Unit and decimals on the graph widget Y-axis.
- Unit and decimals on the singlestat widget.
- MaxWidth option that sets the horizontal scale of the grid.
- Widget grid fixed mode.
- Widget grid adaptive mode.
- Grid implementation for widgets.
- Dynamic X-axis time labels based on time range and steps.
- Templated queries using variables.
- Const and autointerval variables.
- Color override on graph series based on legend regex.
- Templated legends on graph widget.
- Legend on graph widget.
- Graph widget.
- Single metric gather.
- Metric range gather.
- Allow multiple datasources in the same dashboard.
- Debug flag that will set a verbose logger (will break UI rendering but prints errors and infos).
- Termdash render engine implementation for widgets.
- Singlestat widget.
- Gauge widget.
- Main term application.
- Fake metrics gatherer.

[unreleased]: https://github.com/slok/grafterm/compare/v0.2.0...HEAD
[0.2.0]: https://github.com/slok/grafterm/compare/v0.1.0...0.2.0
[0.1.0]: https://github.com/slok/grafterm/releases/tag/v0.1.0


cmd/grafterm/flags.go
package main

import (
	"fmt"
	"os"
	"path"
	"time"

	"github.com/alecthomas/kingpin"
)

// Version is the application version.
var Version = "dev"

// Defaults settings required by flags..
const (
	defConfig          = "dashboard.json"
	defRefreshInterval = "10s"
	defLogPath         = "grafterm.log"
	defGraftermDir     = "grafterm"
)

var defUserDatasourcePath = []string{defGraftermDir, "datasources.json"}

// Env vars.
const (
	envPrefix          = "GRAFTERM"
	envUserDatasources = envPrefix + "_USER_DATASOURCES"
)

// flag descriptions.
const (
	descCfg             = "the path to the configuration file"
	descRefreshInterval = "the interval to refresh the dashboard"
	descLogPath         = "the path where the log output will be written"
	descRelativeDur     = "the relative duration from now to load the graph."
	descStart           = "the time the dashboard will start in time. Accepts 2 formats, relative time from now based on duration(e.g.: 24h, 15m), or fixed duration in ISO 8601 (e.g.: 2019-05-12T09:35:11+00:00). If set it disables relative duration flag."
	descEnd             = "the time the dashboard will end in time. Accepts 2 formats, relative time from now based on duration(e.g.: 24h, 15m), or fixed duration in ISO 8601 (e.g.: 2019-05-12T09:35:11+00:00)."
	descDebug           = "enable debug mode, on debug mode it will print logs to the desired output"
	descVar             = "repeatable flag that will override the variable defined on the dashboard (in 'key=value' form)"
	descDSAlias         = "repeatable flag that maps dashboard ID datasources to user defined datasources in the form of 'dashboard=user' (in 'key=value' form)"
)

var descUserDS = fmt.Sprintf("path to a configuration file with user defined datasources, these datasources can override the dashboard datasources with the same ID and also can be used to alias them using datasource alias flags. It fallbacks to %s env var", envUserDatasources)

type flags struct {
	variables       map[string]string
	aliases         map[string]string
	cfg             string
	userDSPath      string
	debug           bool
	version         bool
	refreshInterval time.Duration
	logPath         string
	start           string
	relativeDur     time.Duration
	end             string
}

func newFlags() (*flags, error) {
	flags := &flags{
		variables: map[string]string{},
		aliases:   map[string]string{},
	}

	// Get default datasource path.
	userHome, _ := os.UserHomeDir()
	userDsPath := ""
	if userHome != "" {
		dsPath := []string{userHome}
		dsPath = append(dsPath, defUserDatasourcePath...)
		userDsPath = path.Join(dsPath...)
	}

	// Create app.
	app := kingpin.New("grafterm", "graph metrics on the terminal")
	app.Version(Version)

	// Register flags.
	app.Flag("cfg", descCfg).Default(defConfig).Short('c').StringVar(&flags.cfg)
	app.Flag("refresh-interval", descRefreshInterval).Default(defRefreshInterval).Short('r').DurationVar(&flags.refreshInterval)
	app.Flag("log-path", descLogPath).Default(defLogPath).StringVar(&flags.logPath)
	app.Flag("relative-duration", descRelativeDur).Short('d').DurationVar(&flags.relativeDur)
	app.Flag("start", descStart).Short('s').StringVar(&flags.start)
	app.Flag("end", descEnd).Short('e').StringVar(&flags.end)
	app.Flag("var", descVar).Short('v').StringMapVar(&flags.variables)
	app.Flag("ds-alias", descDSAlias).Short('a').StringMapVar(&flags.aliases)
	app.Flag("user-datasources", descUserDS).Default(userDsPath).Short('u').Envar(envUserDatasources).StringVar(&flags.userDSPath)
	app.Flag("debug", descDebug).BoolVar(&flags.debug)
	app.Parse(os.Args[1:])

	if err := flags.validate(); err != nil {
		return nil, err
	}

	return flags, nil
}

func (f *flags) validate() error {
	return nil
}
